openedx/core/djangoapps/programs/utils.py

```diff
@@ -345,6 +346,31 @@ def completed_course_runs(self):
         """
         return self.course_runs_with_state['completed']
 
+    def available_date_for_programs(self, program_uuids):
+        """
+        Calculate the available date for completed programs based on
+        course runs.
+
+        Args:
+            program_uuids: a list of uuids for completed programs.
+        """
+        programs = [self._available_date_for_program(program) for program in self.programs if program['uuid'] in program_uuids]
+        return programs
+
+    def _available_date_for_program(self, program):
+        """
+        Calculate the available date for the program based on the courses within it.
+        """
+        program_available_date = datetime.datetime.min.replace(tzinfo=utc)
+        for course in program['courses']:
+            for course_run in course['course_runs']:
+                certificate = certificate_api.get_certificate_for_user(self.user.username, course_run['key'])
```

Comment:
Note for @mikix : This certificate format returns a dict while `available_date_for_certificate` expects `GeneratedCertificate` model object. I was trying to see what we could do here instead of making the direct call to the model, but it doesn't seem like there's much else we can do. It's not a matter of performance, because `get_certificate_for_user` also just makes a query and then turns the cert into a dict.